### PR TITLE
Add package fingerprint

### PR DESCRIPTION
**Pull request by maraino**
### Description

This PR adds the new `fingerprint` package that unifies the logic for X.509 and SSH fingerprint, allowing at the same time supporting other hash functions required by `step`.

### CODE REVIEW

**Comment by maraino** [comment]
@hslatman I've refactor my last PR to better support `step`

File: README.md [hunk 1]
```diff
@@ -75,3 +75,8 @@ or just a YubiKey or an ssh-agent. On the cloud it supports:
 * [Amazon AWS KMS](https://aws.amazon.com/kms/)
 * [Google Cloud Key Management](https://cloud.google.com/security-key-management)
 * [Microsoft Azure Key Vault](https://azure.microsoft.com/en-us/services/key-vault/)
+
+### fingerprint
+
+Package `fingerprint` adds raw support for creating an encoding fingerprints
+using hash functions.
```
Comment (hslatman) [inline README.md:82]:
```suggestion
Package `fingerprint` provides methods for creating and encoding
X.509 certificate, SSH certificate and SSH key fingerprints.
```

File: fingerprint/fingerprint.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	Base64RawURLFingerprint
+	// EmojiFingerprint represents the emoji encoding of the fingerprint.
+	EmojiFingerprint
+)
+
+// New creates a fingerprint of the given data using the hash and returns it
+// encoded with the given encoding format.
+func New(data []byte, h crypto.Hash, encoding Encoding) (string, error) {
+	if !h.Available() {
+		return "", fmt.Errorf("hash function %s is not available", h.String())
```
Comment (hslatman) [inline fingerprint/fingerprint.go:46]:
```suggestion
		return "", fmt.Errorf("hash function %q is not available", h.String())
```

File: fingerprint/fingerprint.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	// This is the default encoding for an SSH key and certificate.
+	Base64RawFingerprint
+	// Base64RawURLFingerprint represents the base64RawURL encoding of the fingerprint.
+	Base64RawURLFingerprint
+	// EmojiFingerprint represents the emoji encoding of the fingerprint.
+	EmojiFingerprint
+)
+
+// New creates a fingerprint of the given data using the hash and returns it
+// encoded with the given encoding format.
```
Comment (hslatman) [inline fingerprint/fingerprint.go:43]:
```suggestion
// New creates a fingerprint of the given data by hashing it and
// returns it in the encoding format.
```

File: fingerprint/fingerprint.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	fp := Fingerprint(hash.Sum(nil), encoding)
+	if fp == "" {
+		return "", fmt.Errorf("unknown encoding value %d", encoding)
+	}
+	return fp, nil
+}
+
+// Fingerprint encoding the given digest using the given encoding format. If an
+// invalid encoding is passed, the return value will be an empty string.
```
Comment (hslatman) [inline fingerprint/fingerprint.go:60]:
```suggestion
// Fingerprint encodes the given digest using the encoding format. If an
// invalid encoding is passed, the return value will be an empty string.
```

File: fingerprint/fingerprint.go (hunk 2 above)
Comment (maraino) [reply fingerprint/fingerprint.go]: Fixed with 3a66062

File: fingerprint/fingerprint.go (hunk 3 above)
Comment (maraino) [reply fingerprint/fingerprint.go]: Fixed with 3a66062

File: fingerprint/fingerprint.go (hunk 4 above)
Comment (maraino) [reply fingerprint/fingerprint.go]: Fixed with 3a66062

**Review by hslatman — APPROVED** [review]

**Pull request merged**